wikidatarefisland/pipes/pipe_one.py

```diff
@@ -0,0 +1,44 @@
+import json
+import jsonlines
+from functools import reduce
+from itertools import chain
+from wikidatarefisland.pipes import statement_filters
+
+
+class PipeOne():
+    def __init__(self, output_file_name, external_id_formatter, blacklisted_properties=[]):
+        self.blacklisted_properties = blacklisted_properties
+        self.output_file_name = output_file_name
+        self.external_id_formatter = external_id_formatter
+
+    def flow(self, input_data):
+        results = map(self.process_item, input_data)
+        with jsonlines.open(self.output_file_name, 'w') as writer:
```

Comment:
I would abstract away the reading the file into Storage class (that already exists) so I can test it with an in-memory mock of storage class and see the result is written.
Another thing: I think this should open the file in "append" mode. We can't write gigabytes of data into the file every time we process the dump. The idea I had was to write to file after checking N items and appending and cleans the variable (basically doing it in batches) it to the file so it doesn't get slowed down by writing to disk non-stop but also the memory usage doesn't get to 100GB (size of json of wikidata is ~960GB)

Reply:
Gotcha!
I actually opted not to use the storage class here because I thought doing this avoided the problem of large in memory objects. 

Rather than trying to read in batches of any defined size I was trying to write python that just used iterators and no complete in memory objects (e.g. lists) but rather a functional style to lazy load / evaluate each step.

Currently both the store and append methods in Storage take a whole in memory object, open a file, write to it, and then close as far as I can see?

Using the the jsonlines builtin (which is just iteratively writing) seems (to me) to open the file and write at the correct rate determined by the system (io.DEFAULT_BUFFER_SIZE?). It also means that we allow python/the os to decide the most appropriate time to flush the buffer and sync the file rather than having to choose and tune a chunk size.

Reply:
It sounds good to use buffer size and jsonlines but I think it should be abstracted away in the Storage class (it would be great if you add it there) so for example it can be used by other classes.
I'm slightly unsure what you mean about leaving this fully to jsonlines and not using append more (maybe I'm missing something obvious, please let me know).
The thing is when we use "with" here, anything inside the clause has its own context and the code makes sure to write the file and flush it away and close the file. That's why we can't use it like this and assume the system batch and write to the file later.

Reply:
Thinking about it when writing my own pipe now I think we should avoid writing to storage altogether, the flow() gets input and returns output. The runner (water facility?) will be responsible for flowing data into the pipe and feeding the output into the next pipe.
This would be way more compliant with SRP too.